File: python/monte_carlo_pi_comparison/monte_carlo_pi.py
import numpy
from mpi4py import MPI
import time
import argparse

real_pi = 3.141592653589793

def monte_carlo_pi(n, seed):
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()
    
    iterations_per_process = n // size

    rng = numpy.random.default_rng(seed * rank)    
    
    count = 0
    for i in range(iterations_per_process):
        x = rng.uniform(-1, 1)
        y = rng.uniform(-1, 1)
        if x * x + y * y < 1:
            count += 1

    count = numpy.array(count, dtype='int')
    count = comm.reduce(count, op=MPI.SUM, root=0)

    if rank == 0:
        return (4 * count / n)
    else:
        return None
    
if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Monte Carlo Pi Calculation')
    parser.add_argument('iterations', type=int, help='Number of iterations')
    args = parser.parse_args()

    try:
        n = int(args.iterations)
        if n <= 0:
            raise ValueError("Number of iterations must be a positive integer.")
    except ValueError as e:
        parser.error(str(e))

    seed = 42
    start_time = time.time()
    pi = monte_carlo_pi(n, seed)
    end_time = time.time()
    if pi is not None:
        print(f"Pi: {pi}, Error: {abs(pi - real_pi)}, Time: {end_time - start_time} seconds")

File: python/monte_carlo_pi_comparison/monte_carlo_pi.cpp
#include <mpi.h>
#include <iostream>
#include <cmath>
#include <cstdlib>
#include <ctime>

const double real_pi = 3.141592653589793;

double monte_carlo_pi(int n, unsigned int* seed) {
    int rank, size, iterations_per_process;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    iterations_per_process = n / size;

    int count = 0;
    *seed *= rank; // Seed for random number generator

    for (int i = 0; i < iterations_per_process; ++i) {
        double x = (double)rand_r(seed) / RAND_MAX * 2.0 - 1.0;
        double y = (double)rand_r(seed) / RAND_MAX * 2.0 - 1.0;
        if (x * x + y * y < 1.0) {
            count++;
        }
    }

    int global_count;
    MPI_Reduce(&count, &global_count, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);

    if (rank == 0) {
        return (4.0 * global_count / (n));
    } else {
        return 0.0;
    }
}

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);

    int n;
    try {
        if (argc != 2 || (n = std::stoi(argv[1])) <= 0) {
            throw std::invalid_argument("Invalid number of iterations");
        }
    } catch (const std::exception& e) {
        int rank;
        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
        if (rank == 0) {
            std::cerr << "Usage: " << argv[0] << " <number_of_iterations>" << std::endl;
            std::cerr << "Error: " << e.what() << std::endl;
            std::cerr << "Please provide a positive integer greater than 0." << std::endl;
        }
        MPI_Finalize();
        return 1;
    }
    unsigned int seed = 42;

    // Start the timer
    double start_time = MPI_Wtime();

    double pi = monte_carlo_pi(n, &seed);

    // Stop the timer
    double end_time = MPI_Wtime();

    int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    if (rank == 0) {
        std::cout << "Pi: " << pi << ", Error: " << fabs(pi - real_pi) << std::endl;
        std::cout << "Execution time: " << end_time - start_time << " seconds" << std::endl;
    }

    MPI_Finalize();
    return 0;
}